// @tanstack/router-plugin@1.99.6/dist/esm/core/router-code-splitter-plugin.js
import { isAbsolute, join, normalize } from "node:path";
import { pathToFileURL, fileURLToPath } from "node:url";
import { logDiff } from "@tanstack/router-utils";
import { getConfig } from "./config.js";
import { compileCodeSplitVirtualRoute, compileCodeSplitReferenceRoute } from "./code-splitter/compilers.js";
import { splitPrefix } from "./constants.js";
const debug = process.env.TSR_VITE_DEBUG && ["true", "router-plugin"].includes(process.env.TSR_VITE_DEBUG);
function capitalizeFirst(str) {
  return str.charAt(0).toUpperCase() + str.slice(1);
}
function fileIsInRoutesDirectory(filePath, routesDirectory) {
  const routesDirectoryPath = isAbsolute(routesDirectory) ? routesDirectory : join(process.cwd(), routesDirectory);
  const path = normalize(filePath);
  return path.startsWith(routesDirectoryPath);
}
const bannedBeforeExternalPlugins = [
  {
    identifier: "@react-refresh",
    pkg: "@vitejs/plugin-react",
    usage: "viteReact()",
    frameworks: ["vite"]
  }
];
class FoundPluginInBeforeCode extends Error {
  constructor(externalPlugin, framework) {
    super(`We detected that the '${externalPlugin.pkg}' was passed before '@tanstack/router-plugin'. Please make sure that '@tanstack/router-plugin' is passed before '${externalPlugin.pkg}' and try again: 
e.g.
plugins: [
  TanStackRouter${capitalizeFirst(framework)}(), // Place this before ${externalPlugin.usage}
  ${externalPlugin.usage},
]
`);
  }
}
const PLUGIN_NAME = "unplugin:router-code-splitter";
const unpluginRouterCodeSplitterFactory = (options = {}, { framework }) => {
  let ROOT = process.cwd();
  let userConfig = options;
  const handleSplittingFile = (code, id) => {
    if (debug) console.info("Splitting Route: ", id);
    const compiledVirtualRoute = compileCodeSplitVirtualRoute({
      code,
      root: ROOT,
      filename: id
    });
    if (debug) {
      logDiff(code, compiledVirtualRoute.code);
      console.log("Output:\n", compiledVirtualRoute.code);
    }
    return compiledVirtualRoute;
  };
  const handleCompilingFile = (code, id) => {
    if (debug) console.info("Compiling Route: ", id);
    const compiledReferenceRoute = compileCodeSplitReferenceRoute({
      code,
      root: ROOT,
      filename: id
    });
    if (debug) {
      logDiff(code, compiledReferenceRoute.code);
      console.log("Output:\n", compiledReferenceRoute.code + "\n\n");
    }
    return compiledReferenceRoute;
  };
  return {
    name: "router-code-splitter-plugin",
    enforce: "pre",
    transform(code, id) {
      if (!userConfig.autoCodeSplitting) {
        return null;
      }
      const url = pathToFileURL(id);
      url.searchParams.delete("v");
      id = fileURLToPath(url).replace(/\\/g, "/");
      if (id.includes(splitPrefix)) {
        return handleSplittingFile(code, id);
      } else if (fileIsInRoutesDirectory(id, userConfig.routesDirectory) && (code.includes("createRoute(") || code.includes("createFileRoute("))) {
        for (const externalPlugin of bannedBeforeExternalPlugins) {
          if (!externalPlugin.frameworks.includes(framework)) {
            continue;
          }
          if (code.includes(externalPlugin.identifier)) {
            throw new FoundPluginInBeforeCode(externalPlugin, framework);
          }
        }
        return handleCompilingFile(code, id);
      }
      return null;
    },
    transformInclude(id) {
      if (!userConfig.autoCodeSplitting) {
        return void 0;
      }
      if (fileIsInRoutesDirectory(id, userConfig.routesDirectory) || id.includes(splitPrefix)) {
        return true;
      }
      return false;
    },
    vite: {
      configResolved(config) {
        ROOT = config.root;
        userConfig = getConfig(options, ROOT);
      }
    },
    rspack(compiler) {
      ROOT = process.cwd();
      userConfig = getConfig(options, ROOT);
    },
    webpack(compiler) {
      ROOT = process.cwd();
      userConfig = getConfig(options, ROOT);
      if (userConfig.autoCodeSplitting && compiler.options.mode === "production") {
        compiler.hooks.done.tap(PLUGIN_NAME, () => {
          console.info("✅ " + PLUGIN_NAME + ": code-splitting done!");
          setTimeout(() => {
            process.exit(0);
          });
        });
      }
    }
  };
};
export {
  unpluginRouterCodeSplitterFactory
};
